core/modules/qdisp/QueryRequest.cc

```diff
@@ -299,20 +302,46 @@ bool QueryRequest::_importStream(JobQuery::Ptr const& jq) {
         // Keep the previous object from wedging the pool.
         _askForResponseDataCmd->notifyFailed();
     }
-    _askForResponseDataCmd = make_shared<AskForResponseDataCmd>(
-                               shared_from_this(), jq, proto::ProtoHeaderWrap::PROTO_HEADER_SIZE);
-    _queueAskForResponse(_askForResponseDataCmd, jq, true);
+
+    // Get the first header from metadata.
+    int expectedLen = proto::ProtoHeaderWrap::PROTO_HEADER_SIZE;
+    int len = expectedLen;
+    const char* buff = GetMetadata(len);
+    if (len != expectedLen) {
+        throw Bug("_importStream wrong header size=" + to_string(len) + " expected=" + to_string(expectedLen));
+    }
+    ResponseHandler::BufPtr bufPtr = make_shared<vector<char>>(len);
```

Comment:
So, now I am confused. It seems that BufPtr is indeed a container. Since a vector carries the length with it, I would suggest not using an external length as you will constantly have to check for consistency.

Reply:
The 2 buffers (this result and next header) are put in one buffer so they can be sent as a single transmit. Once received, the buffer needs to be split into the separate parts so protobufs can decipher them and figure out how big the next bufer should be.